Set a limit on cookie expiration (fixes brave/brave-browser#3443)

fmarier opened the pull request:
The client-side cookie (i.e. document.cookie API) expiry limit is based off of the limit set by both Safari and Firefox:

https://webkit.org/blog/8613/intelligent-tracking-prevention-2-1/
https://groups.google.com/forum/#!msg/mozilla.dev.platform/lECBPeiGTy4/cPP52vyZAwAJ

whereas the server-side cookie (i.e. Set-Cookie header) limit was picked to avoid interfering in a noticeable way with user logins.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

Use the following test pages and examine the state of the cookie jar using the devtools:
- client-side cookies (7-day limit):
  - https://fmarier.github.io/brave-testing/document-cookie.html
- server-side cookies (6-month limit):
  - https://fmarier.org/document-cookie/set-cookie1.html
  - https://fmarier.org/document-cookie/set-cookie2.html
  - https://fmarier.org/document-cookie/set-cookie3.html
  - https://fmarier.org/document-cookie/set-cookie4.html

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

fmarier commented:
@bridiver I forgot to mark this as a draft PR, but basically, at this point, I just want to get feedback on whether or not you think this is a reasonable approach. I will add tests before landing.

darkdh commented on patches/net-cookies-canonical_cookie.cc.patch:
```diff
@@ -0,0 +1,60 @@
+ 
+ }  // namespace
+ 
++const TimeDelta CanonicalCookie::kMaxClientSideExpiration =
```
could you put this into `brave/chromium_src/net/cookies/canonical_cookie.cc` and put it into an anonymous namespace to reduce patch?
Also there is a `TimeDelta::FromDays` that you can use.

darkdh commented on patches/net-cookies-canonical_cookie.cc.patch:
```diff
@@ -0,0 +1,60 @@
+ 
+   DCHECK(!creation_time.is_null());
++  bool is_from_http = !options.exclude_httponly();
+   Time cookie_expires = CanonicalCookie::CanonExpiration(parsed_cookie,
```
we can have an `CanonicalCookie ::BraveCanonExpiration)` calling `CanonicalCookie::CanonExpiration` to get its result and return the min value in `brave/chromium_src/net/cookies/canonical_cookie.cc`.

Also shouldn't we use `ParsedCookie.IsHttpOnly` instead of `CookieOptions.exclude_httponly`?

fmarier replied:
> Also shouldn't we use `ParsedCookie.IsHttpOnly` instead of `CookieOptions.exclude_httponly`?

The naming is a little confusing, but those two functions actually mean something different:
1. `ParsedCookie::IsHttpOnly()` is the `httponly` flag part of the `Set-Cookie` header.
2. `CookieOptions::exclude_httponly()` is a flag set by users of the cookie store to indicate two separate things depending on the context:
   1. whether or not to include `httponly` cookies when doing cookie lookups ("getter path")
   2. whether the cookie creation is happening due to an HTTP header or due to the `document.cookie` DOM API ("setter path")

For this patch, we want to be able to tell the API that the cookie setter is using (i.e. § 2.ii), similar to how `CanonicalCookie::Create()` [uses that flag to reject `httponly` set using `document.cookie`](https://cs.chromium.org/chromium/src/net/cookies/canonical_cookie.cc?l=211-216&rcl=4802de68c80fddacee86ca8427736b006ea1e957).

fmarier commented:
@darkdh I made all of the changes you requested and switched to an override approach.

I also added a unit test which covers the same test cases as my manual test page.
